greeclimate/device.py

```diff
@@ -207,28 +216,44 @@ async def bind(self, key=None):
 
         try:
             if key:
-                self.device_key = key
+                self.device_cipher = cipher_type(key.encode())
             else:
-                await self.send(self.create_bind_message(self.device_info))
-                # Special case, wait for binding to complete so we know that the device is ready
-                task = asyncio.create_task(self.ready.wait())
-                await asyncio.wait_for(task, timeout=self._timeout)
+                if cipher_type is not None:
+                    await self.__bind_internal(cipher_type)
+                else:
+                    """ Try binding with CipherV1 first, if that fails try CipherV2"""
+                    try:
+                        self._logger.info("Attempting to bind to device using CipherV1")
+                        await self.__bind_internal(CipherV1)
+                    except asyncio.TimeoutError:
+                        self._logger.info("Attempting to bind to device using CipherV2")
+                        await self.__bind_internal(CipherV2)
 
         except asyncio.TimeoutError:
             raise DeviceTimeoutError
 
-        if not self.device_key:
+        if not self.device_cipher:
             raise DeviceNotBoundError
         else:
-            self._logger.info("Bound to device using key %s", self.device_key)
+            self._logger.info("Bound to device using key %s", self.device_cipher.key)
+
+    async def __bind_internal(self, cipher_type: type[Union[CipherV1, CipherV2]]):
+        """Internal binding procedure, do not call directly"""
+        default_key = GENERIC_CIPHERS_KEYS.get(cipher_type)
+        await self.send(self.create_bind_message(self.device_info), cipher=cipher_type(default_key))
```

Comment:
May I suggest that you define the generic key as a class attribute inside the CipherV1 and CipherV2 classes. In the `__init__` you can define the key as optional and use the generic key if `key` is None.